File: webApp/packages/elements/src/orchestrator/engine.ts
import type { Effect } from './state';
import type { MethodLifecycleEvent } from './types';
import type { PaymentMethod } from '../shared/types';
import type { OrchestratorCore } from './core';
import type { ResultEvent } from '../shared/types';

export type EffectRunner = {
    run: (effect: Effect) => Promise<void>;
};

export const createOrchestrationEngine = (
    core: OrchestratorCore,
    runner: EffectRunner
) => {
    const dispatchAndRun = async (action: any) => {
        const effects = core.dispatch(action);
        for (const eff of effects) {
            await runner.run(eff);
        }
    };

    return {
        onMethodMounted: (method: PaymentMethod) =>
            dispatchAndRun({ type: 'METHOD_MOUNTED', method }),

        onMethodReady: (method: PaymentMethod) =>
            dispatchAndRun({ type: 'METHOD_READY', method }),

        onLifecycleEvent: (method: PaymentMethod, evt: MethodLifecycleEvent) => {
            if (evt.type === 'SUBMIT_STARTED') {
                return dispatchAndRun({ type: 'METHOD_SUBMIT_STARTED', method });
            }
        },

        onResult: (method: PaymentMethod, result: ResultEvent) =>
            dispatchAndRun({ type: 'METHOD_RESULT', method, result }),

        setActive: (method: PaymentMethod) =>
            dispatchAndRun({ type: 'SET_ACTIVE', method }),
    };
};

File: shared/src/commonMain/kotlin/io/conekta/elements/orchestrator/Engine.kt
package io.conekta.elements.orchestrator

import io.conekta.elements.orchestrator.Effect
import io.conekta.elements.orchestrator.Action
import io.conekta.elements.orchestrator.OrchestratorCore

interface EffectRunner {
    suspend fun run(effect: Effect)
  }
  
class OrchestrationEngine(
    private val core: OrchestratorCore,
    private val runner: EffectRunner
  ) {
    suspend fun dispatch(action: Action) {
        val effects: Array<Effect> = core.dispatch(action)
        for (eff in effects) {
          runner.run(eff)
        }
      }
  }
  